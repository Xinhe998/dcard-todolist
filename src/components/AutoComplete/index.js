import React, { useRef, useState, useContext } from 'react';
import PropTypes from 'prop-types';
import Store from '../../reducers/context';
import * as action from '../../actions/todoActions';
import useClickOutside from '../../hooks/useClickOutside';

import noResult from '../../assets/no_result.png';

import './index.scss';

const AutoComplete = ({
  isOpen, switchHandler, placeHolder, data,
}) => {
  const dispatch = useContext(Store);
  const searchRef = useRef();
  const autocompleteRef = useRef();
  const [searchText, setSearchText] = useState('');
  const searchResult = data.filter(d => d.text.includes(searchText) && searchText !== '');
  useClickOutside(isOpen, autocompleteRef, () => {
    setSearchText('');
    searchRef.current.classList.add('closing');
    setTimeout(() => {
      searchRef.current.classList.remove('closing');
      switchHandler(false);
    }, 300);
  });
  return (
    <div
      className={isOpen ? 'AutoComplete AutoComplete--active' : 'AutoComplete'}
      ref={autocompleteRef}
    >
      <div className="AutoComplete__form-wrapper">
        {isOpen ? (
          <form className="AutoComplete__form">
            <input
              type="text"
              className="AutoComplete__form__input"
              placeholder={placeHolder}
              ref={searchRef}
              onChange={e => setSearchText(e.target.value)}
            />
          </form>
        ) : null}
        <button
          type="button"
          className="AutoComplete__button"
          onClick={() => {
            switchHandler(true);
            setTimeout(() => {
              if (searchRef.current) searchRef.current.focus();
            }, 500);
          }}
        >
          <svg width="30" height="30" viewBox="0 0 36 36">
            <title>Search</title>
            <path
              d="M20.5 22.621l2.146-1.8L28 27.2 25.853 29 20.5 22.621zm-3.479-.618c3.327 0 6.022-2.688 6.022-6.001 0-3.314-2.695-6.002-6.022-6.002C13.695 10 11 12.688 11 16.002c0 3.313 2.695 6.001 6.021 6.001zm0 2c-4.43 0-8.021-3.582-8.021-8.001C9 11.582 12.591 8 17.021 8c4.43 0 8.022 3.582 8.022 8.002 0 4.419-3.592 8.001-8.022 8.001z"
              fillRule="nonzero"
              fill="#d8d8d8"
            />
          </svg>
        </button>
      </div>
      {isOpen && searchText ? (
        <div className="AutoComplete__result-wrapper">
          <ul className="AutoComplete__result-list">
            {searchResult.length ? (
              searchResult.map(item => (
                <li
                  key={item.id}
                  onClick={() => {
                    dispatch(action.setTaskShowingDetail({ id: item.id }));
                    switchHandler(false);
                    setSearchText('');
                  }}
                >
                  {item.text}
                </li>
              ))
            ) : (
              <div className="AutoComplete__result__no-result-wrapper">
                <img className="AutoComplete__result__no-result-img" src={noResult} alt="no search result." />
                <p className="AutoComplete__result__no-result-info">No Result</p>
              </div>
            )}
          </ul>
        </div>
      ) : null}
    </div>
  );
};

AutoComplete.propTypes = {
  isOpen: PropTypes.bool,
  switchHandler: PropTypes.func,
  placeHolder: PropTypes.string,
  data: PropTypes.array,
};
AutoComplete.defaultProps = {
  isOpen: false,
  switchHandler: null,
  placeHolder: '',
  data: [],
};
export default AutoComplete;
