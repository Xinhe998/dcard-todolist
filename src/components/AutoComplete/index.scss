.AutoComplete {
  // width: 300px;
  max-width: 500px;
  position: relative;
  margin-left: auto;
  margin-right: 0px;
  right: 25px;
  // transition: all .5s;
  &.AutoComplete--active {
    margin-right: 80px;
    right: 0;
    .AutoComplete__button {
      position: absolute;
      top: 0px;
      right: -60px;
    }
  }
  .AutoComplete__button {
    background: none;
    border: none;
    color: var(--secondary-color-gray1);
    outline: 0;
    cursor: pointer;
    float: right;
    border-radius: 50px;
    width: 36px;
    height: 36px;
    padding: 5px 3px;
    text-align: center;
    transition: all 0.2s;
    &:hover {
      background: #005a8d;
    }
  }
}

.AutoComplete__form-wrapper {
  display: flex;
  flex-direction: row-reverse;
  justify-content: flex-end;
}
.AutoComplete__form {
  width: 100%;
}
.AutoComplete__form__input {
  width: 0%;
  padding: 5px 50px 5px 10px;
  background: none;
  border: none;
  font-size: 14px;
  font-weight: 400;
  height: 28px;
  border-radius: 10px;
  background: #005a8d;
  outline: 0;
  color: #ffffff;
  animation: searchField_stretch 0.3s, searchField_shodowShow 0.3s 0.5s ease-in;
  animation-direction: alternate;
  animation-fill-mode: forwards;
  transition: all 0.3s ease-in;
  direction: ltr;
  &.closing {
    animation: searchField_closing 0.5s;
  }
  &::placeholder {
    color: var(--secondary-color-gray2);
    opacity: 0.4;
  }
  &:focus {
    box-shadow: 0 0 0 5px rgba(0, 0, 0, 0.08);
  }
}

.AutoComplete__result-wrapper {
  width: 270px;
  border-radius: 10px;
  background: #ffffff;
  box-shadow: var(--card-box-shadow);
  color: var(--text-color);
  position: absolute;
  padding: 0;
  margin-top: 10px;
  .AutoComplete__result-list {
    list-style: none;
    padding: 0;
    li {
      cursor: pointer;
      padding: 10px 30px;
      &:hover {
        background: var(--secondary-color-gray2);
      }
    }
  }
}

@keyframes searchField_stretch {
  from {
    width: 0%;
  }
  to {
    width: 100%;
  }
}

@keyframes searchField_shodowShow {
  0% {
    box-shadow: none;
  }
  70% {
    box-shadow: 0 0 0 3px rgba(0, 0, 0, 0.08);
  }
  100% {
    box-shadow: 0 0 0 5px rgba(0, 0, 0, 0.08);
  }
}

@keyframes searchField_closing {
  from {
    width: 100%;
  }
  to {
    width: 0%;
  }
}