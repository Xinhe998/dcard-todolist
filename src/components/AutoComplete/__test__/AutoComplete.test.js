/* eslint-disable func-names */
import 'babel-polyfill';
import React from 'react';
import { shallow, configure, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import AutoComplete from '../index';

import filter from '../../../assets/filter.png';

// 以該解析器提供給 Enzyme 做渲染 Component 的設置
configure({ adapter: new Adapter() });

describe('<AutoComplete />', function () {
  let testComp;
  beforeEach(() => {
    jest.resetModules();
    this.params = {
      isOpen: true,
      switchHandler: jest.fn(),
      placeHolder: '',
      data: null,
    };

    this.makeSubject = () => {
      const { isOpen, switchHandler, placeHolder, data } = this.params;

      testComp = shallow(
        <AutoComplete
          isOpen={isOpen}
          switchHandler={switchHandler}
          placeHolder={placeHolder}
          data={data}
        />,
      );
      return testComp;
    };
  });
  describe('when it has complete data', () => {
    beforeEach(() => {
      this.params.placeHolder = 'Test...';
      this.subject = this.makeSubject();
    });
    it('should be same as snapshot', async () => {
      const wrapper = render(await testComp);
      expect(wrapper).toMatchSnapshot();
    });
    it('should be three options', async () => {
      expect(
        testComp.find('.AutoComplete__tooltip__list__option'),
      ).toHaveLength(3);
    });
    // 測試icon
    it('should render icon', () => {
      expect(testComp.find('img').prop('src')).toEqual(filter);
    });
    // 測試text
    it('should render text', () => {
      expect(testComp.find('.AutoComplete__button').text()).toEqual(
        'Test:option1',
      );
    });
    // 測試按下 button 會觸發 swichOptionHandler
    it('test click event', () => {
      testComp.find('button').simulate('click');
      expect(this.params.swichOptionHandler.mock.calls.length).toEqual(1);
    });
    // 測試按下選項
    it('test click event', () => {
      testComp
        .find('.AutoComplete__tooltip__list__option')
        .at(0)
        .simulate('click');
      expect(this.params.swichOptionHandler.mock.calls.length).toEqual(1);
      expect(this.params.handleClickDispatch.mock.calls.length).toEqual(1);
    });
  });
});
