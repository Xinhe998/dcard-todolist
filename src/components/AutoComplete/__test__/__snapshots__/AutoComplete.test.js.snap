// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoComplete /> when it has complete data should be same as snapshot 1`] = `
initialize {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
